framework/include/base/DependencyResolverInterface.h

```diff
@@ -72,6 +90,9 @@ DependencyResolverInterface::sort(typename std::vector<T> & vector)
 
   for (typename std::vector<T>::iterator iter = start; iter != end ; ++iter)
   {
+    if ((*iter)->allowCyclicDepenedency())
+      continue;
```

Comment:
I don't think I like any of these changes regarding the DependencyResolveInterface. 

Consider all of the changes this caused.
- You're adding `validParams` to this interface where there was none before. 
- Changing the constructor to accept a Parameter object.
- Changing several classes that inherit from this interface to provide Parameters
- Augmenting all of those interfaces with a parameter that shouldn't be set by most of them
- Using that parameter to bypass the entire purpose of this class
- Still sorting based on potentially incomplete or bad information (what is the behavior of sorting with cyclic dependancies)

This is simply the wrong design here.

Perhaps the DiscreteMaterial shouldn't participate or use the DependencyResolver Interface, or if you DO need it (you might), maybe it shouldn't use the same Resolver because it is allowed to have cyclic dependencies. Basically if it isn't sorting correctly, why is it in here at all?


You're adding Parameters to the interface so you can access a specific set a Boolean to basically turn off the whole